analyzers/src/SonarAnalyzer.Common/Helpers/MessageTemplates.cs

```diff
@@ -0,0 +1,64 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Text.RegularExpressions;
+
+namespace SonarAnalyzer.Helpers;
+
+/// <summary>
+/// Grammar can be found <a href="https://github.com/messagetemplates/grammar"> here</a>.
+/// </summary>
+public static class MessageTemplates
+{
+    private const string NamePattern = "[0-9a-zA-Z_]+";
+    private const string PlaceholderPattern = $"(?<Placeholder>{NamePattern})";
+    private const string AlignmentPattern = "(,-?[0-9]+)?";
+    private const string FormatPattern = @"(:[^\}]+)?";
+
+    private const string HolePattern = "{" + "[@$]?" + PlaceholderPattern + AlignmentPattern + FormatPattern + "}";
+    private const string TextPattern = @"([^\{]|\{\{|\}\})+";
+    private const string TemplatePattern = $"^({TextPattern}|{HolePattern})*$";
+
+    private static readonly Regex TemplateRegex = new(TemplatePattern, RegexOptions.Compiled, TimeSpan.FromMilliseconds(300));
```

Comment:
We will have to be careful when validating performance. 300ms can add up pretty fast.

Reply:
We [plan to raise](https://trello.com/c/vfvMIRbq/1243-rules-with-regex-use-safeismatch-everywhere) the default regex to around 300ms.

Also, from testing it seems like in the happy path it's much faster, it only becomes slower when there is a syntax error. That's because in the happy path there is almost no backtracking, which is not true when there are syntax errors.

I think I prefer to leave it slightly higher than necessary, so that the issues are not flaky on a CI run with a not-so-great machine.